tests/HTMLPurifier/Lexer/Html5LexTest.php

```diff
@@ -0,0 +1,748 @@
+    {
+        $this->assertTokenization(
+            '<b><<</b>',
+            array(
+                new HTMLPurifier_Token_Start('b'),
+                new HTMLPurifier_Token_Text('<<'),
+                new HTMLPurifier_Token_End('b'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize eon1996.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_eon1996()
+    {
+        $this->assertTokenization(
+            '< <b>test</b>',
+            array(
+                new HTMLPurifier_Token_Text('< '),
+                new HTMLPurifier_Token_Start('b'),
+                new HTMLPurifier_Token_Text('test'),
+                new HTMLPurifier_Token_End('b'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize a body tag inside CDATA.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_bodyInCDATA()
+    {
+        $this->assertTokenization(
+            '<![CDATA[<body>Foo</body>]]>',
+            array(
+                new HTMLPurifier_Token_Text('<body>Foo</body>'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize a valid HTML segment.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_()
+    {
+        $this->assertTokenization(
+            '<a><img /></a>',
+            array(
+                new HTMLPurifier_Token_Start('a'),
+                new HTMLPurifier_Token_Empty('img'),
+                new HTMLPurifier_Token_End('a'),
+            )
+        );
+    }
+
+    /**
+     * Tokenize a HTML conditional comment.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_ignoreIECondComment()
+    {
+        $this->assertTokenization(
+            '<!--[if IE]>foo<a>bar<!-- baz --><![endif]-->',
+            array()
+        );
+    }
+
+    /**
+     * Tokenize an XML processing instruction.
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_removeProcessingInstruction()
+    {
+        $this->config->set('Core.RemoveProcessingInstructions', true);
+
+        $this->assertTokenization(
+            '<?xml blah blah ?>',
+            array()
+        );
+    }
+
+    /**
+     * Tokenize plain text with a mixture of new line characters (CR / LF).
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_removeNewline()
+    {
+        $this->config->set('Core.NormalizeNewlines', true);
+        $this->assertTokenization(
+            "plain\rtext\r\n",
+            array(
+                new HTMLPurifier_Token_Text("plain\ntext\n")
+            )
+        );
+    }
+
+    /**
+     * Tokenize plain text with a mixture of new line characters (CR / LF).
+     *
+     * @return void
+     */
+    public function test_tokenizeHTML_noRemoveNewline()
+    {
+        $this->config->set('Core.NormalizeNewlines', false);
+
+        $this->assertTokenization(
+            "plain\rtext\r\n",
+            array(
+                new HTMLPurifier_Token_Text("plain\rtext\r\n")
```

Comment:
masterminds/html5 normalises `\r\n` to `\n` so `Core.NormalizeNewlines` is useless when this lexer is used.

https://github.com/Masterminds/html5-php/blob/54d066b1664a5fb749511c764e8d208f6bc80c3b/src/HTML5/Parser/Scanner.php#L336

Reply:
Logged https://github.com/Masterminds/html5-php/issues/176

Reply:
Updated the test. Probably need to document this in the README?